File: clients/macos/SyncBridgeMac/Core/LocalSend/LocalSendModels.swift
// LocalSendModels.swift

import Foundation

struct LocalPeer: Identifiable, Equatable {
    let id: String
    let name: String
    let platform: String
    let host: String
    let port: Int
}

enum LocalTransferPhase: String {
    case idle, connecting, waitingAccept, transferring, paused, completed, failed
}

enum LocalTransferDirection { case sending, receiving }

struct LocalFileProgress: Identifiable {
    let index: Int
    let name: String
    let size: Int64
    var transferred: Int64

    var id: Int { index }
    var percent: Double { size > 0 ? Double(transferred) / Double(size) : 0 }
    var remaining: Int64 { max(0, size - transferred) }
}

struct LocalTransferProgress: Identifiable {
    let transferId: String
    let peerName: String
    let direction: LocalTransferDirection
    var phase: LocalTransferPhase
    var files: [LocalFileProgress]
    var speedBytesPerSec: Int64 = 0
    var error: String?

    var id: String { transferId }
}

struct IncomingLocalOffer: Identifiable {
    let offer: LocalSendProtocol.Offer
    let peer: LocalPeer
    var id: String { offer.id }
}

File: clients/android/app/src/main/java/com/syncbridge/android/localsend/LocalSendModels.kt
package com.syncbridge.android.localsend

data class LocalPeer(
    val id: String,
    val name: String,
    val platform: String,
    val host: String,
    val port: Int,
)

enum class LocalTransferPhase {
    Idle,
    Connecting,
    WaitingAccept,
    Transferring,
    Paused,
    Completed,
    Failed,
}

data class LocalTransferProgress(
    val transferId: String,
    val peerName: String,
    val direction: LocalTransferDirection,
    val phase: LocalTransferPhase,
    val files: List<LocalFileProgress>,
    val speedBytesPerSec: Long = 0,
    val error: String? = null,
)

enum class LocalTransferDirection { Sending, Receiving }

data class LocalFileProgress(
    val index: Int,
    val name: String,
    val size: Long,
    val transferred: Long,
) {
    val percent: Float
        get() = if (size <= 0) 0f else (transferred.toFloat() / size.toFloat()).coerceIn(0f, 1f)

    val remaining: Long get() = (size - transferred).coerceAtLeast(0)
}

data class IncomingOffer(
    val offer: LocalSendProtocol.Offer,
    val peer: LocalPeer,
)
